api/controllers/ListUserController.js

```diff
@@ -272,57 +328,11 @@ module.exports = {
     // Synchronize google spreadsheets
     promises.push(GSSSyncService.deleteUserFromSpreadsheets(list._id, user.id));
     promises.push(OutlookService.deleteUserFromContactFolders(list._id, user.id));
+    logger.info(
+      '[ListUserController->checkout] Saving list and user for checkout and sending notifications',
+    );
     await Promise.all(promises);
```

Comment:
About these Promises executions: I see we're adding all the logging as various conditions are detected, but it's kind of premature. The actual code which might succeed or fail is accumulated into this `Promise.all`, which has no `then`/`catch` to detect whether something worked or not.

I assume if one of them fails, it 500s the whole chain of events, right? Even though some or many might have already succeeded. That seems to be the behavior I've seen when I debug locally and have some minor config problem (like a missing GSSSync key). It reports failures when nothing critical actually failed.

Maybe it would be better to queue these logging events and only execute them once `Promise.all` succeeds:

```js
const pendingLogs = []; // push the log message each time you push to `promises`
pendingLogs.push({type: 'info', messages: ['something was successful', object]});
pendingLogs.push({type: 'info', messages: ['something else was successful', string]});

await Promise.all(promises).then((res) => {
  pendingLogs.forEach(log => {
    logger[log.type](...log.messages);
  });
}).catch((err) => {
  logger.error('[ListUserController->update] some operations failed during batch execution', err);
});
```

(this is technically pseudocode, it might not work. just to get the idea across)

Reply:
I agree when you say that logging when adding the promise to the promise array is probably premature, but at the same time I believe that once all logging messages have been added, if an error happens in, say, GSSSyncService, it will be logged.
Also, in the pseudocode you wrote, the only difference I see with what I'm proposing is that you're writing the log messages after everything gets executed, while I'm writing them before things get executed, and I don't really understand why this would be more helpful for debugging...

Reply:
By logging after `Promise.all` succeeds, your logs are verified to be reporting events which actually occurred.

If you log beforehand and your `Promise.all` fails (since it can fail when just one single Promise throws an exception), then any number of things that got logged beforehand might not have actually worked.

I suppose if we take async out of the picture, logging beforehand is like doing this with jQuery:

```js
// premature log before event ever happens
console.log('the document is ready');

$document.ready(function () {
  // the console.log statements should be HERE inside the callback,
  // where we know the document is ready
  //
  // using callback is equivalent to .then() in Promises or async/await
  $('.something').thatRequiresDomReady();
});

// it is definitely not ready here either
console.log('the document is definitely ready now');
```

Reply:
If you log afterwards, a lot of things could also happen and succeed before you actually get any log, and you won't know what happened because nothing got logged...
If I look back at your pseudocode and assume there is an error so we go into the "catch" block, you will see the error message in the logs, but you won't be able to tell which promises actually tried to execute (and eventually succeeded) before you got to the error message.
So basically, I think either way, the fact that we are executing in an async manner can make logs relatively messy, because one execution won't lead to the same log trace as another one...

Reply:
To sum up our post-standup chat today (myself, @emmajane, @guillaumev), queuing the logs and playing them back inside `Promise.all().then()` has the advantage of avoiding log output during times when known validation errors occur (typically a `logger.warn` followed by `throw Boom`), which would avoid the confusion of logging prematurely, but never actually executing the planned operation.


Additionally, we can add the output of `pendingLogs` to the `.catch()` block as well, in order to indicate what did (or didn't) happen before exceptions that occur during `Promise.all`.